Translation for the page 'API Reference -> DOM Elements' (#127)
reactjs/it.reactjs.org: content/docs/reference-dom-elements.md

  - "tips/dangerously-set-inner-html.html"
---

React implementa un sistema DOM indipendente dal browser per massimizzare le prestazioni e la compatibilità con i vari browsers. Abbiamo colto l'occasione per smussare gli angoli di alcune implementazioni nel DOM del browser.

> In React, tutte le proprietà DOM e gli attributi (inclusi i gestori degli eventi) dovrebbero essere [_camelCased_](https://it.wikipedia.org/wiki/Notazione_a_cammello). Ad esempio, l'attributo HTML `tabindex` corrisponde all'attributo `tabIndex` in React. Fanno eccezione gli attributi `aria-*` e `data-*`, i quali dovrebbero essere in minuscolo. Poi mantenere `aria-label` come `aria-label` per esempio.

Reviewer: ```suggestion
In React, tutte le proprietà DOM e gli attributi (inclusi i gestori degli eventi) dovrebbero essere [_camelCased_](https://it.wikipedia.org/wiki/Notazione_a_cammello). Ad esempio, l'attributo HTML `tabindex` corrisponde all'attributo `tabIndex` in React. Fanno eccezione gli attributi `aria-*` e `data-*`, i quali dovrebbero essere in minuscolo. Puoi mantenere `aria-label` come `aria-label` per esempio.
```

## Differenze Negli Attributi {#differences-in-attributes}

Ci sono alcuni attributi che funzionano in modo diverso tra React ed HTML:

### checked {#checked}

L'attributo `checked` è supportato dai componenti `<input>` di tipo `checkbox` o `radio`. Puoi usarlo per impostare il fatto che il componente sia appunto _checked_ ovvero selezionato/attivato. Ci torna utile nella costruzione di componenti controllati. `defaultChecked` è l'equivalente non controllato, il quale imposta l'attributo `checked` del componente la prima volta che esso viene montato.

### className {#classname}

Per specificate una classe CSS, utilizza l'attributo `className`. Si applica a tutti i normali elementi DOM ed SVG come `<div>`, `<a>` ed altri.

> Se stai utilizzando React con i [Web Components](https://developer.mozilla.org/en-US/docs/Web/Web_Components) (fatto poco comune), usa l'attributo `class` invece.

Reviewer: ```suggestion
Se stai utilizzando React con i [Web Components](https://developer.mozilla.org/en-US/docs/Web/Web_Components) (fatto poco comune), usa invece l'attributo `class`.
```

### dangerouslySetInnerHTML {#dangerouslysetinnerhtml}

`dangerouslySetInnerHTML` è la controparte React all'uso di `innerHTML` nel browser DOM. In generale, impostare HTML dal codice è rischioso in quanto può esporre facilmente i tuoi utenti ad attacchi [cross-site scripting (XSS)](https://it.wikipedia.org/wiki/Cross-site_scripting). Puoi impostare HTML direttamente da React, ma devi farlo utilizzando `dangerouslySetInnerHTML` e passando un oggetto avente chiave `__html` per ricordarti del fatto che è pericoloso: Ad esempio:

```js
function creaCodiceHtml() {
  return {__html: 'Primo &middot; Secondo'};
}

function MioComponente() {
  return <div dangerouslySetInnerHTML={creaCodiceHtml()} />;
}
```

### htmlFor {#htmlfor}

Dato che `for` è una parola riservata in JavaScript, gli elementi React utilizzano `htmlFor`.

### onChange {#onchange}

> L'evento `onChange` si comporta così come di consueto: qualndo un campo in un form cambia, l'evento viene lanciato. Non utilizziamo intenzionalmente il comportamento esistente nel browser perché il nome `onChange` non descrive molto bene il suo comportamento e React si basa su questo evento per gestire gli input dell'utente in tempo reale.

Reviewer: ```suggestion
L'evento `onChange` si comporta così come di consueto: quando un campo in un form cambia, l'evento viene lanciato. Non utilizziamo intenzionalmente il comportamento esistente nel browser perché il nome `onChange` non descrive molto bene il suo comportamento e React si basa su questo evento per gestire gli input dell'utente in tempo reale.
```

### selected {#selected}

> L'attributo `selected` è supportato dai componenti `<option>`. Puoi utilizzarlo quando devi far si che un componente sia selezionato. Questo attributo ci torna utile anche nella costruzione di componenti controllati.

Reviewer: ```suggestion
L'attributo `selected` è supportato dai componenti `<option>`. Puoi utilizzarlo quando devi far sì che un componente sia selezionato. Questo attributo ci torna utile anche nella costruzione di componenti controllati.
```

### style {#style}

>Nota
>
>Alcuni esempi nella documentazione fanno uso di `style` per comodità, tuttavia **l'utilizzo dell'attributo `style` come metodo principale per la personalizzazione degli stili degli elementi non è generalmente raccomandato.** Nella maggioranza dei casi, [`className`](#classname) dovrebbe essere usato per far riferimento a classi definite in fogli di stile CSS esterni. `style` viene usato maggiormente nelle applicazioni React per aggiungere stili calcolati dinamicamente durante la renderizzazione. Vedi anche [FAQ: Stili e CSS](/docs/faq-styling.html).

> L'attributo `style` accetta un oggetto JavaScript con proprietà [_camelCased_](https://it.wikipedia.org/wiki/Notazione_a_cammello) invece di una stringa CSS. Parallelamente alla proprietà `style` del DOM JavaScript, è più efficiente e previene vulnerabilità XSS. Ad esempop:

Reviewer: ```suggestion
L'attributo `style` accetta un oggetto JavaScript con proprietà [_camelCased_](https://it.wikipedia.org/wiki/Notazione_a_cammello) invece di una stringa CSS. Proprio come la proprietà `style` del DOM JavaScript, è più efficiente e previene vulnerabilità XSS. Ad esempio:
```

```js
const divStile = {
  color: 'blue',
  backgroundImage: 'url(' + imgUrl + ')',
};

function ComponenteCiaoMondo() {
  return <div style={divStile}>Ciao Mondo!</div>;
}
```

Nota come gli stili non vengono prefissati automaticamente. Per supportare browser più vecchi, dovrai aggiungere le rispettive proprietà:

```js
const divStile = {
  WebkitTransition: 'all', // nota la lettera maiuscola 'W'
  msTransition: 'all' // 'ms' è l'unico prefisso vendor (terza parte) in minuscolo
};

function ComponenteConTransizione() {
  return <div style={divStile}>Questo dovrebbe funzionare in tutti i browser</div>;
}
```

> Le chiavi di stile sono in notazione a cammello in modo da essere consistenti con l'accesso alle proprietà dei nodi DOM da JavaScript (ad esempio `node.style.backgroundImage`). Prefissi vendor (terze parti) [diversi da `ms`](https://www.andismith.com/blogs/2012/02/modernizr-prefixed/) dovrebbero iniziare con la lettera maiuscola. Ecco perché `WebkitTransition` inizia con "W".

Reviewer: ```suggestion
Le chiavi di stile sono in notazione camel case in modo da essere consistenti con l'accesso alle proprietà dei nodi DOM da JavaScript (ad esempio `node.style.backgroundImage`). I prefissi vendor (terze parti) [diversi da `ms`](https://www.andismith.com/blogs/2012/02/modernizr-prefixed/) dovrebbero iniziare con la lettera maiuscola. Ecco perché `WebkitTransition` inizia con "W".
```

React aggiunge automaticamente un suffisso "px" ad alcune proprietà di stile in formato numerico. Se vorrai utilizzare unità diverse al posto di "px", dovrai specificare il valore come stringa insieme all'unità desiderata. Per esempio:

```js
// Risultato: '10px'
<div style={{ height: 10 }}>
  Ciao Mondo!
</div>

// Risultato: '10%'
<div style={{ height: '10%' }}>
  Ciao Mondo!
</div>
```

> Non tutte proprietà di stile vengono convertite in stringhe. Alcune di esse restano senza unità (ad esempio `zoom`, `order`, `flex`). Una lista completa delle proprietà senza unità è disponibile [qui](https://github.com/facebook/react/blob/4131af3e4bf52f3a003537ec95a1655147c81270/src/renderers/dom/shared/CSSProperty.js#L15-L59).

Reviewer: ```suggestion
Non tutte le proprietà di stile vengono convertite in stringhe. Alcune di esse restano senza unità (ad esempio `zoom`, `order`, `flex`). Una lista completa delle proprietà senza unità è disponibile [qui](https://github.com/facebook/react/blob/4131af3e4bf52f3a003537ec95a1655147c81270/src/renderers/dom/shared/CSSProperty.js#L15-L59).
```

### suppressContentEditableWarning {#suppresscontenteditablewarning}

Normalmente, si riceve un messaggio di avviso quando un elemento con figli viene anche marcato come `contentEditable` in quanto non funziona. Questo attributo sopprime tale messaggio di avviso. Non utilizzarlo a meno che tu stia costruendo una libreria come [Draft.js](https://facebook.github.io/draft-js/) che gestisce i `contentEditable` manualmente.

### suppressHydrationWarning {#suppresshydrationwarning}

> Se utilizzi la renderizzazione lato server di React, normalmente si ha un avviso quando il server ed il cliente renderizzano contenuti differenti. Comunque, in alcuni rari casi, è molto difficile, se non impossibile, garantire una totale corrispondenza tra i contenuti. Ad esempio, i [timestamps](https://it.wikipedia.org/wiki/Marca_temporale) saranno certamente differenti tra server e client.

Reviewer: ```suggestion
Se utilizzi la renderizzazione lato server di React, normalmente si ha un avviso quando il server ed il client renderizzano contenuti differenti. Comunque, in alcuni rari casi, è molto difficile se non impossibile garantire una totale corrispondenza tra i contenuti. Ad esempio, i [timestamps](https://it.wikipedia.org/wiki/Marca_temporale) saranno certamente differenti tra server e client.
```

> Se imposti `suppressHydrationWarning` a `true`, React non avviserà più in caso di differenze negli attributi e nel contenuto dell'elemento.funziona solamente ad un livello di profondità ed è un modo per aggirare il problema. Non abusarne. Puoi trovare maggiori dettagli riguardo l'_hydration_ nella [documentazione di `ReactDOM.hydrate()`](/docs/react-dom.html#hydrate).

Reviewer: ```suggestion
Se imposti `suppressHydrationWarning` a `true`, React non avviserà più in caso di differenze negli attributi e nel contenuto dell'elemento. Questa tecnica funziona solamente per il primo livello di profondità ed è un modo per aggirare il problema. Non abusarne. Puoi trovare maggiori dettagli riguardo l'_hydration_ nella [documentazione di `ReactDOM.hydrate()`](/docs/react-dom.html#hydrate).
```

### value {#value}

L'attributo `value` è supportato dai componenti `<input>` e `<textarea>`. Puoi utilizzarlo per impostare il valore dei componenti. Torna utile nella costruzione di componenti controllati. `defaultValue` è la controparte non controllata, la quale imposta il valore del componente quando viene montato per la prima volta.

## Tutti gli Attributi HTML Supportati {#all-supported-html-attributes}

Fino a React 16, tutti gli attributi DOM standard [o custom](/blog/2017/09/08/dom-attributes-in-react-16.html) sono pienamente supportati.

> React ha sempre offerto delle API JavaScript-centriche per il DOM. Dato che i componenti React utilizzano spesso sia props custom che props relative al DOM, React usa la `notazione a cammello` come convenzione così come avviene con le API del DOM:

Reviewer: ```suggestion
React ha sempre offerto delle API JavaScript-centriche per il DOM. Dato che i componenti React utilizzano spesso sia props custom che props relative al DOM, React usa la notazione camel case  ("notazione a cammello") come convenzione così come avviene con le API del DOM:
```
Everywhere in Italy, as far as I can tell, people just say "camel case". Should we add it to the untranslatable terms list in the Glossary?
